### PR TITLE
refactor(react-chart): migrate react chart to TS 

### PR DESCRIPTION
**Pull request by Krijovnick**

### CODE REVIEW

File: packages/dx-react-chart/src/templates/pattern.tsx [hunk 1]
```diff
@@ -1,7 +1,14 @@
 import * as React from 'react';
-import * as PropTypes from 'prop-types';
+import { PatternProps } from '../types';
 
-export class Pattern extends React.PureComponent {
+const defaultProps = {
+  size: 6,
+  opacity: 0.75,
+};
+type PatternDefaultProps = Readonly<typeof defaultProps> ;
```
Comment (ushkal) [inline packages/dx-react-chart/src/templates/pattern.tsx]: Unnecessary whitespace

File: packages/dx-chart-core/src/plugins/animation/computeds.ts [hunk 2]
```diff
@@ -41,7 +41,7 @@ const getPieAnimationName = () => {
 
 const getDefaultAreaAnimationOptions = () => '1s';
 
-const getDefaultPieAnimationOptions = ({ index }: Point) => `${0.7 + index * 0.1}s`;
+const getDefaultPieAnimationOptions = ({ index }: {index: number}) => `${0.7 + index * 0.1}s`;
```
Comment (DmitryBogomolov) [inline packages/dx-chart-core/src/plugins/animation/computeds.ts]: Why is it changed?

File: packages/dx-chart-core/src/plugins/series/computeds.ts [hunk 3]
```diff
@@ -97,12 +97,12 @@ getBarPointTransformer.isBroad = true;
...
 export const findSeriesByName = (
-  name: string, series: SeriesList,
-) => series.find(seriesItem => seriesItem.symbolName === name);
+  name: symbol, series: SeriesList,
+): Series => series.find(seriesItem => seriesItem.symbolName === name) as Series;
 
 export const dBar = ({
   x, y, y1, width,
-}: TransformedPoint & { width: number }) => ({
+}: { x: number, y: number, y1: number, width: number }) => ({
```
Comment (DmitryBogomolov) [inline packages/dx-chart-core/src/plugins/series/computeds.ts]: Why is it changed?

File: packages/dx-chart-core/src/types/chart-core.types.ts [hunk 4]
```diff
@@ -102,6 +102,9 @@ export interface Series {
   readonly valueField: string;
   readonly createHitTester: CreateHitTesterFn;
   readonly getPointTransformer: GetPointTransformerFn;
+  readonly pointComponent: React.ComponentType<any>;
+  readonly state: string;
```
Comment (DmitryBogomolov) [inline packages/dx-chart-core/src/types/chart-core.types.ts]: By its meaning it should be `state?`. How does it work?

File: packages/dx-chart-core/src/types/plugins.animation.types.ts [hunk 5]
```diff
@@ -1,13 +1,16 @@
 import { PureComputed } from '@devexpress/dx-core';
 import {
-  Scale, Point,
+  Scale,
 } from './chart-core.types';
 export type Scales = {
   readonly xScale: Scale;
   readonly yScale: Scale;
 };
+type Point = {index: number};
```
Comment (DmitryBogomolov) [inline packages/dx-chart-core/src/types/plugins.animation.types.ts]: Why it replaces original `Point`?

File: packages/dx-chart-core/src/utils/event-tracker.ts [hunk 6]
```diff
@@ -39,7 +40,7 @@ const buildEventHandler = (seriesList: SeriesList, handlers: HandlerFnList): Eve
     hitTesters = hitTesters || createHitTesters();
     const targets: TrackerTarget[] = [];
     seriesList.forEach(({ name: series, index: order, symbolName }) => {
-      const status = hitTesters![symbolName](location);
+      const status = hitTesters![symbolName as unknown as string](location);
```
Comment (DmitryBogomolov) [inline packages/dx-chart-core/src/utils/event-tracker.ts]: Looks like `string` should be replaced with `symbol` in `HitTesters` declartion.

File: packages/dx-react-chart/src/chart.tsx [hunk 7]
```diff
@@ -16,7 +16,15 @@ import { Palette } from './plugins/palette';
...
 import { Label } from './templates/label';
 
-class RawChart extends React.PureComponent {
+const defaultProps = {
+  height: 500,
+};
+
+type RawChartDefaultProps = Readonly<typeof defaultProps>;
+
+class RawChart extends React.PureComponent<RawChartProps & RawChartDefaultProps> {
```
Comment (DmitryBogomolov) [inline packages/dx-react-chart/src/chart.tsx]: It should be `.PureComponent<RawChartProps>`.
`RawChartDefaultProps` is a subset of `RawChartProps` and should already be contained in the latter.

File: packages/dx-react-chart/src/chart.tsx [hunk 8]
```diff
@@ -16,7 +16,15 @@ import { Palette } from './plugins/palette';
...
+  static defaultProps = defaultProps;
```
Comment (DmitryBogomolov) [inline packages/dx-react-chart/src/chart.tsx]: Try the following (maybe there is no need to have separate `defaultProps` constant):
```js
static defaultProps = {
  height: 500,
};
```

File: packages/dx-react-chart/src/chart.tsx [hunk 9]
```diff
@@ -57,23 +66,10 @@ class RawChart extends React.PureComponent {
...
-  children: PropTypes.node,
-};
-
-RawChart.defaultProps = {
-  height: 500,
-  width: undefined,
-  children: null,
-};
 
 RawChart.components = {
```
Comment (DmitryBogomolov) [inline packages/dx-react-chart/src/chart.tsx]: Initialize this field together with declaration.
```js
static component: PluginComponents = {
  rootComponent: 'Root',
};
```

File: packages/dx-react-chart/src/plugins/axis.tsx [hunk 10]
```diff
@@ -33,7 +34,20 @@ const adjustScaleRange = (scale, [width, height]) => {
...
+  tickSize: 5,
+  indentFromAxis: 10,
+};
+type RawAxisDefaultProps = Readonly<typeof defaultProps>;
+
+class RawAxis extends React.PureComponent<RawAxisProps & RawAxisDefaultProps> {
+  static components: PluginComponents;
+  static defaultProps = defaultProps;
+
+  rootRef: React.RefObject<HTMLElement>;
```
Comment (DmitryBogomolov) [inline packages/dx-react-chart/src/plugins/axis.tsx]: With *class fields* available `constructor` can be removed in some components.
```js
adjustedWidth = 0;
adjustedHeight = 0;
rootRef = React.createRef();
```

File: packages/dx-react-chart/src/plugins/axis.tsx [hunk 11]
```diff
@@ -102,12 +116,12 @@ class RawAxis extends React.PureComponent {
...
                     flexGrow: dx || undefined,
                   }}
-                  ref={this.rootRef}
+                  ref={this.rootRef as any}
                 >
                   <svg
                     width={this.adjustedWidth}
                     height={this.adjustedHeight}
-                    style={SVG_STYLE}
+                    style={SVG_STYLE as any}
```
Comment (DmitryBogomolov) [inline packages/dx-react-chart/src/plugins/axis.tsx]: What kind of error is shown without `as any`?

File: packages/dx-react-chart/src/plugins/basic-data.tsx [hunk 12]
```diff
@@ -1,9 +1,8 @@
 import * as React from 'react';
-import * as PropTypes from 'prop-types';
 import { Plugin, Getter } from '@devexpress/dx-react-core';
-import { defaultDomains } from '@devexpress/dx-chart-core';
+import { defaultDomains, DataItems } from '@devexpress/dx-chart-core';
 
-export const BasicData = ({ data }) => (
+export const BasicData = ({ data }: { data: DataItems }) => (
```
Comment (DmitryBogomolov) [inline packages/dx-react-chart/src/plugins/basic-data.tsx]: It should something like `BasicData: React.SFC<BasicDataProps> = ({ data }) => { ... }`.

File: packages/dx-react-chart/src/types/templates.axis.types.ts [hunk 13]
```diff
@@ -0,0 +1,37 @@
+export interface LabelProps {
+  // The label text
+  text: string | number;
```
Comment (DmitryBogomolov) [inline packages/dx-react-chart/src/types/templates.axis.types.ts]: Why not just `string`?

File: packages/dx-react-chart/src/types/templates.axis.types.ts [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+  y2: number;
+}
+
+export interface RootProps {
+  // The x coordinate of the top left corner of the axis’ rendering area
+  dx: number;
+  // The y coordinate of the top left corner of the series’ rendering area
+  dy: number;
+  /** @internal */
+  onSizeChange: any;
```
Comment (DmitryBogomolov) [inline packages/dx-react-chart/src/types/templates.axis.types.ts]: Why `any`? It is declared in *dx-react-core* - `SizerProps.onSizeChange`.

File: packages/dx-react-chart/src/types/templates.series.types.ts [hunk 15]
```diff
@@ -0,0 +1,110 @@
+import {
+  Scales, BuildAnimatedStyleGetterFn, TransformedPoint,
+} from '@devexpress/dx-chart-core';
+import { PureComputed } from '@devexpress/dx-core';
+
+type Path = PureComputed<[TransformedPoint[]], string>;
```
Comment (DmitryBogomolov) [inline packages/dx-react-chart/src/types/templates.series.types.ts]: It does not relate to computeds. Declare it as
```js
type Path = (points: ReadonlyArray<TransformedPoint>) => string
```
Maybe it should even be named `PathFn` instead of `Path`, though I am not sure of it.

File: packages/dx-react-chart/src/types/templates.series.types.ts [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+  /** @internal */
+  scales: Scales;
+  /** @internal */
+  getAnimatedStyle: BuildAnimatedStyleGetterFn;
+  /** @internal */
+}
+
+interface InternalPointProps {
+/** @internal */
+  argument: any;
```
Comment (DmitryBogomolov) [inline packages/dx-react-chart/src/types/templates.series.types.ts]: Why is `value` field absent?

File: packages/dx-react-chart/src/plugins/axis.tsx [hunk 17]
```diff
@@ -20,7 +21,7 @@ const SVG_STYLE = {
   position: 'absolute', left: 0, top: 0, overflow: 'visible',
 };
 
-const adjustScaleRange = (scale, [width, height]) => {
+const adjustScaleRange = (scale: Scale, [width, height]: [number, number]) => {
```
Comment (DmitryBogomolov) [inline packages/dx-react-chart/src/plugins/axis.tsx]: There is `Size` type in *dx-react-core*.

File: packages/dx-react-core/src/types.ts [hunk 18]
```diff
@@ -6,8 +6,19 @@ export type StateHelper = {
...
 
 export type CreateStateHelperFn = PureComputed<
-  [React.PureComponent, { [fieldName: string]: () => ActionFn<any> | undefined }],
+  [React.PureComponent, { [fieldName: string]: () => ActionFn<any> | undefined }?],
   StateHelper
 >;
 
 export type ActionFn<P> = PureComputed<[P], void | boolean>;
+export type Size = {
+  width: number;
```
Comment (DmitryBogomolov) [inline packages/dx-react-core/src/types.ts]: Add `readonly` modifier.

File: packages/dx-react-core/src/types.ts [hunk 19]
```diff
@@ -6,8 +6,19 @@ export type StateHelper = {
...
+  height: number;
+};
+
+export type SizerProps = {
+  onSizeChange: (size: Size) => void;
+  onScroll?: (e) => void;
```
Comment (DmitryBogomolov) [inline packages/dx-react-core/src/types.ts]: Isn't there a warning the `e` type is not explicitly defined?

File: packages/dx-react-chart/src/types/templates.series.types.ts [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+
+interface InternalPathProps {
+  /** @internal */
+  index: number;
+  /** @internal */
+  state?: string;
+  /** @internal */
+  style?: any;
+  /** @internal */
+  scales: Scales;
```
Comment (DmitryBogomolov) [inline packages/dx-react-chart/src/types/templates.series.types.ts]: `scales`, `state`, `getAnimatedStyle` should not be declared in several interfaces - there should be one (like the code that makes these fields in *series-helper*.

Something like
```js
interface <Connected>XXX {
  index: number;
  state?: string;
  scales: Scales;
  getAnimatedStyle: GetAnimatedStyleFn;
}

interface <Connected>Series extends <Connected>XXX {
  pointComponent: ...
}

interface <Connected>Point extends <Connected>XXX {
  seriesIndex: number;
}
```

File: packages/dx-chart-core/src/plugins/series/computeds.ts (hunk 3 above)
Comment (Krijovnick) [reply packages/dx-chart-core/src/plugins/series/computeds.ts]: Because there is ts error: missing the extra properties from type 'TransformedPoint': argument, value, index, color.

File: packages/dx-chart-core/src/plugins/animation/computeds.ts (hunk 2 above)
Comment (Krijovnick) [reply packages/dx-chart-core/src/plugins/animation/computeds.ts]: There is ts error: missing the extra properties from type `Point` argument, value, color.

Comment (DmitryBogomolov) [reply packages/dx-chart-core/src/plugins/animation/computeds.ts]: It has to be fixed some other way then as function argument is actual of `TransformedPoint` type.

File: packages/dx-chart-core/src/plugins/series/computeds.ts (hunk 3 above)
Comment (DmitryBogomolov) [reply packages/dx-chart-core/src/plugins/series/computeds.ts]: Argument is of `<Connected>BarPoint` type. `TransformedPoint` is the most base type with a required set of properties.

File: packages/dx-chart-core/src/utils/event-tracker.ts (hunk 6 above)
Comment (Krijovnick) [reply packages/dx-chart-core/src/utils/event-tracker.ts]: There is ts error: Type 'symbol' cannot be used as an index type.

File: packages/dx-react-chart/src/chart.tsx (hunk 7 above)
Comment (Krijovnick) [reply packages/dx-react-chart/src/chart.tsx]: There are several variants. If I use  `.PureComponent<RawChartProps>` then I should add `!` to the `height` property in the code. Sometimes it brings a lot of trouble.  The hack with `RawChartDefaultProps` solve that problem. Look the [article](https://medium.com/@martin_hotell/react-typescript-and-defaultprops-dilemma-ca7f81c661c7) where I have find that possibility.

File: packages/dx-react-chart/src/plugins/axis.tsx (hunk 11 above)
Comment (Krijovnick) [reply packages/dx-react-chart/src/plugins/axis.tsx]: `Type '{ position: string; left: number; top: number; overflow: string; }' is not assignable to type 'CSSProperties'.
  Types of property 'position' are incompatible.
    Type 'string' is not assignable to type '"absolute" | "-moz-initial" | "inherit" | "initial" | "revert" | "unset" | "-webkit-sticky" | "fixed" | "relative" | "static" | "sticky" | undefined'`

File: packages/dx-react-chart/src/types/templates.axis.types.ts (hunk 14 above)
Comment (Krijovnick) [reply packages/dx-react-chart/src/types/templates.axis.types.ts]: I'll fix it

File: packages/dx-react-chart/src/types/templates.axis.types.ts (hunk 13 above)
Comment (Krijovnick) [reply packages/dx-react-chart/src/types/templates.axis.types.ts]: I get it from our [documentation](https://devexpress.github.io/devextreme-reactive/react/chart/docs/reference/value-axis/). I think it is because `tick` can be formatted or returned as it is (number): https://github.com/DevExpress/devextreme-reactive/blob/master/packages/dx-chart-core/src/plugins/axis/computeds.ts#L18

File: packages/dx-react-chart/src/types/templates.series.types.ts (hunk 16 above)
Comment (Krijovnick) [reply packages/dx-react-chart/src/types/templates.series.types.ts]: It is internal properties. This interface extends RawPointProps, RawSliceProps and RawBarProps interfaces and they have `value` prop.

File: packages/dx-react-core/src/types.ts (hunk 19 above)
Comment (Krijovnick) [reply packages/dx-react-core/src/types.ts]: No, there is no warning. I think because it is `any` and no inconsistencies in types.

File: packages/dx-react-chart/src/plugins/axis.tsx (hunk 17 above)
Comment (Krijovnick) [reply packages/dx-react-chart/src/plugins/axis.tsx]: The `Size` in `dx-react-core` is object, not the array

Comment (Krijovnick) [reply packages/dx-react-chart/src/plugins/axis.tsx]: But I have find several places were such array using, thus I'll create extra type for that.

File: packages/dx-react-chart/src/chart.tsx (hunk 7 above)
Comment (ushkal) [reply packages/dx-react-chart/src/chart.tsx]: At the moment, it works because withComponents returns 'any' type. Once it fixed, this will not work because of incompatible props types: 
`export const Chart: React.ComponentType<RawChartProps> = withComponents({ Root })(RawChart);`

File: packages/dx-react-chart/src/plugins/axis.tsx (hunk 11 above)
Comment (ushkal) [reply packages/dx-react-chart/src/plugins/axis.tsx]: SVG_STYLE should be typed as style: 
`const SVG_STYLE: React.CSSProperties - ...`

File: packages/dx-chart-core/src/types/chart-core.types.ts (hunk 4 above)
Comment (Krijovnick) [reply packages/dx-chart-core/src/types/chart-core.types.ts]: The properties was added for avoid ts errors in [that](https://github.com/DevExpress/devextreme-reactive/blob/master/packages/dx-react-chart/src/utils/series-helper.jsx#L53) code. It is work because `seriesComponent` is any.

File: packages/dx-react-chart/rollup.config.js [hunk 21]
```diff
@@ -1,24 +1,30 @@
...
   ],
   external: external(__dirname),
   plugins: [
-    resolve({
-      main: false,
-      extensions: ['.js', '.jsx'],
+    typescriptRollup({
+      typescript,
+      useTsconfigDeclarationDir: true,
+      abortOnError: false,
```
Comment (ushkal) [inline packages/dx-react-chart/rollup.config.js]: This need to be removed in prod

File: packages/dx-react-chart/src/plugins/pane-layout.tsx [hunk 22]
```diff
@@ -13,28 +12,24 @@ import {
...
   flex: 1, zIndex: 1, position: 'relative', width: '100%',
 };
 
-const SVG_STYLE = {
+const SVG_STYLE: React.CSSProperties = {
   position: 'absolute', left: 0, top: 0, overflow: 'visible',
 };
 
-const SizerContainer = ({ children }) => (
+type SizerContainerProps = {
```
Comment (ushkal) [inline packages/dx-react-chart/src/plugins/pane-layout.tsx]: It's unnecessary - props contain children by default

File: packages/dx-react-chart/src/plugins/selection-state.tsx [hunk 23]
```diff
@@ -0,0 +1,22 @@
+import * as React from 'react';
+import {
+  Plugin,
+  Getter,
+  Getters,
+} from '@devexpress/dx-react-core';
+import { changeSeriesState, SELECTED } from '@devexpress/dx-chart-core';
+import { SelectionStateProps } from '../types';
+
+/** @internal */
```
Comment (ushkal) [inline packages/dx-react-chart/src/plugins/selection-state.tsx]: Isn't this plugin public? :)

File: packages/dx-react-chart/src/plugins/stack.tsx [hunk 24]
```diff
@@ -0,0 +1,32 @@
+import * as React from 'react';
+import { Plugin, Getter, Getters } from '@devexpress/dx-react-core';
+import { getStackedSeries } from '@devexpress/dx-chart-core';
+import { StackProps, StacksOptions, OffsetFn, OrderFn } from '../types';
+import {
+  stackOrderNone,
+  stackOffsetDiverging,
+} from 'd3-shape';
+
+/** @internal */
```
Comment (ushkal) [inline packages/dx-react-chart/src/plugins/stack.tsx]: Stack is also public

File: packages/dx-chart-core/src/types/chart-core.types.ts [hunk 25]
```diff
@@ -4,137 +4,141 @@ export type DataItems = ReadonlyArray<DataItem>;
...
+  /** @internal */
+  readonly symbolName: unique symbol;
 }
+/** @internal */
 export type SeriesList = ReadonlyArray<Series>;
 
+/** @internal */
 export type GetValueDomainFn = (points: PointList) => DomainItems;
 
 export type Palette = ReadonlyArray<string>;
```
Comment (ushkal) [inline packages/dx-chart-core/src/types/chart-core.types.ts]: Palette is also a plugin, we should export that plugin and rename this type

File: packages/dx-react-chart/src/plugins/title.tsx [hunk 26]
```diff
@@ -1,21 +1,30 @@
...
 
 import {
   Plugin,
   Template,
   TemplatePlaceholder,
+  PluginComponents,
 } from '@devexpress/dx-react-core';
+import { TitleProps } from '../types';
+
+/** @internal */
```
Comment (ushkal) [inline packages/dx-react-chart/src/plugins/title.tsx]: It's public too

File: packages/dx-react-chart/src/types/index.ts [hunk 27]
```diff
@@ -0,0 +1,39 @@
...
+
+  AreaSeries, LineSeries, SplineSeries, BarSeries, ScatterSeries, PieSeries,
+  SeriesProps, PathComponentProps, PathComponentPathProps, PointComponentProps,
+  AreaSeriesProps,
+  LineSeriesProps,
+  SplineSeriesProps,
+  BarSeriesProps,
+  ScatterSeriesProps,
+  PieSeriesProps,
+// } from '@devexpress/dx-chart-core';
```
Comment (ushkal) [inline packages/dx-react-chart/src/types/index.ts]: Redundant comment

File: packages/dx-chart-core/src/types/chart-core.types.ts (hunk 25 above)
Comment (Krijovnick) [reply packages/dx-chart-core/src/types/chart-core.types.ts]: Done. Palette -> Colors
